### DISCUSSION TITLE
We got DMCA'd again!

### DISCUSSION THREAD

**Discussion by ZXMushroom63** [EaglerReborn/support#5] [Announcements]
Hello, EaglerReborn has once again been DMCA'd.

The reason being: "Their code goes against our EULA and Terms of Use by bypassing server-based authentication and changing memory state on the client to simulate a successful sign in without actually performing any form of authentication or license check with the authentication server."

This _is_ a valid DMCA reason, unfortunately.

What is going to happen is still being decided. At any rate, please notify us if you have the latest repo (I only have the PluginAPI a5.3 commit)

Bye for now,

ZXMushroom63

**Comment by wxnnvs** [comment]
I think i got the EaglerPL repo from a5.2

**Comment by radmanplays** [comment]
@ZXMushroom63 i think i have the latest repo but its missing the commit that made the title screen say "beta 1.2" but the version of plugin api in the repo i have is a5.3 : 
![image](https://github.com/EaglerReborn/support/assets/95340057/d4bd6dd9-cab5-40fa-8b33-43a35e98c0bf)
 
i uploaded it to this repo : https://github.com/Eaglercraft-Archive/Eaglerreborn-beta-v1.2-src

**Comment by leahanderson1** [comment]
forgot to say that i uploaded my 1.2 source

**Comment by radmanplays** [comment]
@ZXMushroom63 will you come back?

**Reply by ZXMushroom63** [reply]
idk, should I?
I've got a lot of other stuff going on, and Eaglercraft as a whole feels pretty dead.
I don't think there is really a way to work on EaglerReborn legally, and I've lost the motivation by now.

Every new repo we will make will get DMCA'd just as fast. Mojang has a legal team, we don't.

I don't see what there is to come back to, but it was fun while it lasted.

**Reply by ZXMushroom63** [reply]
Also, forgot to mention this, I don't think Java server support works at *all* anymore. Repl.it has put everything, even static deployments, behind a paywall.

**Reply by radmanplays** [reply]
> idk, should I?
> I've got a lot of other stuff going on, and Eaglercraft as a whole feels pretty dead.
> I don't think there is really a way to work on EaglerReborn legally, and I've lost the motivation by now.
> 
> Every new repo we will make will get DMCA'd just as fast. Mojang has a legal team, we don't.
> 
> I don't see what there is to come back to, but it was fun while it lasted.

but eaglercraft u21 has added singleplayer to the game and eaglercraft now has a [dmca protected, self hosted git](https://git.eaglercraft.online/eaglercraft) and since @OtterCodes101 is friends with lax1dude he could ask lax1dude to make a repo for eaglerreborn and give access to it to you and otter

**Reply by leahanderson1** [reply]
> Also, forgot to mention this, I don't think Java server support works at _all_ anymore. Repl.it has put everything, even static deployments, behind a paywall.

it doesn't, i'm working on a new relay soon

**Reply by leahanderson1** [reply]
> but eaglercraft u21 has added singleplayer to the game and eaglercraft now has a [dmca protected, self hosted git](https://git.eaglercraft.online/eaglercraft) and since @OtterCodes101 is friends with lax1dude he could ask lax1dude to make a repo for eaglerreborn and give access to it to you and otter

the repo you are referring to is maintained by nuraad and the rest of the eagler discord admins, but i could ask

**Reply by radmanplays** [reply]
@ZXMushroom63 https://github.com/eaglerforge

**Reply by ZXMushroom63** [reply]
Cool :D

Btw, you could try something pretty cool to convert most of the PluginAPI in one go, if you want to try. It uses `Proxy`, which allows you to write dynamic access methods for a non-modified exposed Java object.

**Reply by radmanplays** [reply]
@ZXMushroom63 ive successfully ported all of eaglerreborn to eaglerforge and the only thing left is the network global but when i ported it and tried to join a singleplayer world(eaglercraft now has singleplayer btw) the world wouldnt load so i opened the dev console and found like 10000 errors in the dev console. after a while i found [this issue](https://github.com/EaglerReborn/support/issues/7) in the support repo meaning this bug has existed since the network global was added to eaglerreborn but we didnt know since singleplayer didnt exist yet. the errors suggest that things like "server_ip" is undefined in packetjoin, etc and idk how to fix it

**Reply by ZXMushroom63** [reply]
That means that when you tried to join the world, the packetjoin event tried to get the IP and failed. I don't know enough about the issue, but try to not set the server IP on the packet data.

**Reply by ZXMushroom63** [reply]
Also, with the previous issue, I still don't know what causes it. Probably me failing to handle some scenario.

**Reply by leahanderson1** [reply]
also @ZXMushroom63 check out what someone made using radman's fork of the PluginAPI
![image](https://github.com/EaglerReborn/support/assets/144633084/426df49b-eb1c-457f-8516-3076a6795ed8)
[null_-_wzzz_-_2024-03-08_06-40-55.webm](https://github.com/EaglerReborn/support/assets/144633084/88b27b2e-fe99-40e9-ab57-ca7d5de56939)
i think https://github.com/Murturtle made this

**Reply by ZXMushroom63** [reply]
Cool! Is that just a GUI change on top of @radmanplays's fork, or also other features as well? Also is that singleplayer?

**Reply by ZXMushroom63** [reply]
Wait, the GUI is changed with a plugin. Lol, that's clever

**Reply by ZXMushroom63** [reply]
@radmanplays I recommend scrapping the entire plugin/mod API. It was designed badly, requiring manually porting every single class and property. I saw you have a repo of Eaglercraft that uses TeaVM 0.9, which properly supports exposing objects to JavaScript. This means that the only manual it that you need to do is events and classes, and the rest can be done as a JS library that hooks into exposed objects and classes. (eg: instead of doing strange things with packets, I can get the packet's name and send an event containing the packet, all via overriding functions in JS.)

**Reply by radmanplays** [reply]
@ZXMushroom63 can you give me a example js library with `displayToChat({msg: String})` ? also eaglercraft has been using teavm 9.0 since u20 (the singleplayer update) and the repo i have was made before u20 (its u19)

**Reply by radmanplays** [reply]
@ZXMushroom63

**Reply by wxnnvs** [reply]
bro is ghosting him

**Reply by ZXMushroom63** [reply]
Sorry! I have don't check my notifications very often :/
With display to chat, given you exposed constructor function proxies and the Minecraft instance to JavaScript, you could write:
```javascript
var mc = window.Minecraft; //Exposed mc object
//Also assuming a ChatComponentText constructor proxy function is exposed to window as well
window.PluginAPI = {};
PluginAPI.displayToChat = function (t) {
  mc.ingameGUI.getChatGUI().printChatMessage(ChatComponentText(t));
}
```

**Reply by radmanplays** [reply]
@ZXMushroom63  do i need unminified classes.js for it to work and even if i used unminified wouldnt the functions and variables in the exposed mc object start with $ ? like `PluginAPI.javaclient.$currentScreen` ?

**Reply by radmanplays** [reply]
@ZXMushroom63 would `setGlobal("javaclient", mc);` work or do i have to use teavm to somehow expose it?

**Reply by ZXMushroom63** [reply]
It depends whether you're using TeaVM 0.9 in your repo or not.
If you are, you can use `setGlobal("javaclient", mc);` without any worries.

If you're not though, you need to do a bit of javascript proxying to make it work.
```javascript
class TeaVMFixerProxy extends Proxy {
  constructor(object) {
     this.object = object;
  }
  get(target, prop, receiver) {
    var data = object["$"+prop];
    if (Object.prototype.toString.call(data) === "[object Object]") {
      return new TeaVMFixerProxy(data);
    }
    return data;
  },
}
```
Modifying setGlobal to push params through this proxy would remove the $ prefix.

**Reply by radmanplays** [reply]
@ZXMushroom63 i use TeaVM 0.9.2 and the $ prefix is still there. also the file size gets too large when i try to compile a unminified classes.js
